cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py

```diff
@@ -129,63 +113,51 @@ def notify_user_converging_status(kube_dns):
 
 
-@when('sdn-plugin.available', 'docker.available')
-@when_not('sdn.configured')
-def container_sdn_setup(sdn):
-    ''' Receive the information from the SDN plugin, and render the docker
-    engine options. '''
-    hookenv.status_set('maintenance',
-                       'Configuring container runtime with SDN.')
-    sdn_config = sdn.get_sdn_config()
-
-    opts = DockerOpts()
-    opts.add('bip', sdn_config['subnet'])
-    opts.add('mtu', sdn_config['mtu'])
-
-    with open('/etc/default/docker', 'w') as stream:
-        stream.write('DOCKER_OPTS="{}"'.format(opts.to_s()))
-    _reconfigure_docker_for_sdn()
-    set_state('sdn.configured')
-
-
 @when('kubernetes-worker.components.installed', 'kube-api-endpoint.available',
       'tls_client.ca.saved', 'tls_client.client.certificate.saved',
       'tls_client.client.key.saved')
 @when_not('kube-dns.available')
 def start_worker(kube_api):
-    '''We have related to either an api server or a load balancer connected
-    to the apiserver along with the certificate and keys. Render the init
-    scripts.'''
-    create_config(kube_api)
-    render_init_scripts(kube_api)
-    restart_unit_services()
+    '''Need to start the worker services before the api-server can schedule
+    required addons like DNS inside the workers.'''
+    # Get the list of kubernetes api servers from the relationship object.
+    servers = get_kube_api_servers(kube_api)
+    if data_changed('kube-api-servers', servers):
+        create_config(servers[0])
+        render_init_scripts(servers)
+        restart_unit_services()
+        update_kubelet_status()
 
 
 @when('kubernetes-worker.components.installed', 'kube-api-endpoint.available',
       'tls_client.ca.saved', 'tls_client.client.certificate.saved',
       'tls_client.client.key.saved', 'kube-dns.available')
 def render_dns_scripts(kube_api, kube_dns):
     ''' The dns is now available, re-render init config with DNS data. '''
+    # GEt the list of kubernetes api servers from the relationship object.
+    servers = get_kube_api_servers(kube_api)
     # Fetch the DNS data on the relationship.
     dns = kube_dns.details()
-    # Initialize a FlagManager object to add flags to unit data.
-    opts = FlagManager('kubelet')
-    # Append the DNS flags + data to the FlagManager object.
-    opts.add('--cluster-dns', '{0}:{1}'.format(dns['sdn-ip'], dns['port']))
-    opts.add('--cluster-domain', dns['domain'])
-
-    create_config(kube_api)
-    set_state('kubernetes-worker.config.created')
-    render_init_scripts(kube_api)
-    restart_unit_services()
+    if (data_changed('kube-api-servers', servers) or
+            data_changed('kube-dns', dns)):
+        # Initialize a FlagManager object to add flags to unit data.
+        opts = FlagManager('kubelet')
+        # Append the DNS flags + data to the FlagManager object.
+        opts.add('--cluster-dns', '{0}:{1}'.format(dns['sdn-ip'], dns['port']))
```

Comment:
While this does not have to be corrected right away, we do need either a method invocation on the flag manager to do an upstert on the contained data, or we will need to pop the existing clusterdns flags off the dictionary and re-add the new values.

This will cause us some grief if the SDN adapter updates its CIDR, TTL's expire and it runs on a different subnet, et-al. We'll have a dead link DNS in the flags, and a proper DNS IP in the flags. The question is which one will win and what kind of impact will that have on performance?